File: app/src/main/java/com/billpayments/features/auth/data/AuthRepository.kt
package com.billpayments.features.auth.data

import com.billpayments.core.mock.MockNetwork

class InvalidOtpException : Exception()

interface AuthRepository {
    suspend fun sendOtp(phone: String)
    suspend fun verifyOtp(phone: String, code: String): String
}

class FakeAuthRepository(private val network: MockNetwork) : AuthRepository {

    override suspend fun sendOtp(phone: String) = network.delay()

    override suspend fun verifyOtp(phone: String, code: String): String {
        network.delay()
        val isSixDigits = code.length == 6 && code.toIntOrNull() != null
        if (!isSixDigits) throw InvalidOtpException()
        return "user-$phone"
    }
}

File: BillPayments/Features/Auth/Data/AuthRepository.swift
struct InvalidOtpError: Error {}

protocol AuthRepository {
    func sendOtp(phone: String) async throws
    func verifyOtp(phone: String, code: String) async throws -> String
}

final class FakeAuthRepository: AuthRepository {
    private let network: MockNetwork

    init(network: MockNetwork) {
        self.network = network
    }

    func sendOtp(phone: String) async throws {
        try await network.delay()
    }

    func verifyOtp(phone: String, code: String) async throws -> String {
        try await network.delay()
        let isSixDigits = code.count == 6 && Int(code) != nil
        guard isSixDigits else { throw InvalidOtpError() }
        return "user-\(phone)"
    }
}
